File: lcof/of013/Solution.py
# 深度搜索
class Solution:
    def movingCount(self, m: int, n: int, k: int) -> int:
        def bitSum(x):
            s = 0
            while x != 0:
                s += x % 10
                x //= 10
            return s

        def dfs(i, j, si, sj):
            if i >= m or j >= n or k < si + sj or (i, j) in visited: return 0
            visited.add((i, j))
            return 1 + dfs(i + 1, j, bitSum(i + 1), sj) + dfs(i, j + 1, si, bitSum(j + 1))

        visited = set()
        return dfs(0, 0, 0, 0)


class Solution2:
    def movingCount(self, m: int, n: int, k: int) -> int:
        def bitSum(x):
            s = 0
            while x != 0:
                s += x % 10
                x //= 10
            return s

        queue, visited = [(0, 0, 0, 0)], set()

        while queue:
            i, j, si, sj = queue.pop()
            if i >= m or j >= n or k < si + sj or (i, j) in visited: continue
            visited.add((i, j))
            queue.append((i + 1, j, bitSum(i + 1), sj))
            queue.append((i, j + 1, si, bitSum(j + 1)))

        return len(visited)


if __name__ == '__main__':
    print(Solution().movingCount(2, 3, 1))
    print(Solution().movingCount(38, 15, 9))

    print(Solution2().movingCount(2, 3, 1))
    print(Solution2().movingCount(38, 15, 9))

File: lcof/of013/Solution.go
package Solution

import "container/list"

func movingCount1(m int, n int, k int) int {
	queue := list.List{}
	visited := make([][]bool, m)
	for i := 0; i < len(visited); i++ {
		visited[i] = make([]bool, n)
	}
	queue.PushBack([]int{0, 0, 0, 0})
	res := 0
	for queue.Len() > 0 {
		back := queue.Back()
		queue.Remove(back)
		bfs := back.Value.([]int)
		i := bfs[0]
		j := bfs[1]
		si := bfs[2]
		sj := bfs[3]
		if i >= m || j >= n || si+sj > k || visited[i][j] {
			continue
		}

		res++
		visited[i][j] = true

		sj1 := sj + 1
		si1 := si + 1
		if (j+1)%10 == 0 {
			sj1 = sj - 8
		}
		if (i+1)%10 == 0 {
			si1 = si - 8
		}

		queue.PushBack([]int{i + 1, j, si1, sj})
		queue.PushBack([]int{i, j + 1, si, sj1})
	}
	return res
}

var n1, m1, k1 int
var visited [][]bool

func movingCount2(m int, n int, k int) int {
	m1 = m
	n1 = n
	k1 = k
	visited = make([][]bool, m)
	for i := 0; i < len(visited); i++ {
		visited[i] = make([]bool, n)
	}

	return dfs(0, 0, 0, 0)
}

func dfs(i int, j int, si int, sj int) int {
	if i >= m1 || j >= n1 || si+sj > k1 || visited[i][j] {
		return 0
	}
	visited[i][j] = true

	sj1 := sj + 1
	si1 := si + 1
	if (j+1)%10 == 0 {
		sj1 = sj - 8
	}
	if (i+1)%10 == 0 {
		si1 = si - 8
	}

	return 1 + dfs(i, j+1, si, sj1) + dfs(i+1, j, si1, sj)
}
